crates/pepper/src/provider.rs

```diff
@@ -0,0 +1,454 @@
+/// directory. In a future optimization, this could be enhanced to manage multiple
+/// `ListingTables` (one per virtual file) and union their results for better control
+/// over file-level operations.
+pub struct PepperTableProvider {
+    /// Table metadata from the catalog
+    table_metadata: TableMetadata,
+    /// Reference to the metadata catalog for file operations
+    _catalog: Arc<dyn MetadataCatalog>,
+    /// Underlying Vortex `ListingTable` that scans all virtual files in the table directory
+    /// Note: Each `DataFile` in the catalog represents a subdirectory (virtual file),
+    /// but this `ListingTable` currently scans all of them together
+    listing_table: Arc<ListingTable>,
+}
+
+impl std::fmt::Debug for PepperTableProvider {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("PepperTableProvider")
+            .field("table_metadata", &self.table_metadata)
+            .finish_non_exhaustive()
+    }
+}
+
+impl PepperTableProvider {
+    /// Create a new Pepper table provider.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the table cannot be found in the catalog or if the listing
+    /// table cannot be created.
+    pub async fn new(table_name: &str, catalog: Arc<dyn MetadataCatalog>) -> CatalogResult<Self> {
+        let table_metadata = catalog.get_table(table_name).await?;
+
+        // Create listing table for the Vortex files
+        let table_path = &table_metadata.path;
+        let dir_url_str = if table_path.ends_with('/') {
+            table_path.to_string()
+        } else {
+            format!("{table_path}/")
+        };
+
+        let table_url = ListingTableUrl::parse(&dir_url_str).map_err(|e| {
+            super::catalog::CatalogError::InvalidOperation {
+                message: e.to_string(),
+            }
+        })?;
+
+        let format = Arc::new(VortexFormat::default());
+        let listing_options = ListingOptions::new(format);
+
+        let config = ListingTableConfig::new(table_url)
+            .with_listing_options(listing_options)
+            .with_schema(Arc::<arrow_schema::Schema>::clone(&table_metadata.schema));
+
+        let listing_table = ListingTable::try_new(config).map_err(|e| {
+            super::catalog::CatalogError::InvalidOperation {
+                message: e.to_string(),
+            }
+        })?;
+
+        Ok(Self {
+            table_metadata,
+            _catalog: catalog,
+            listing_table: Arc::new(listing_table),
+        })
+    }
+
+    /// Create a new table in Pepper.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the table cannot be created in the catalog.
+    pub async fn create_table(
+        catalog: Arc<dyn MetadataCatalog>,
+        options: CreateTableOptions,
+    ) -> CatalogResult<Self> {
+        let _table_id = catalog.create_table(options.clone()).await?;
+        Self::new(&options.table_name, catalog).await
+    }
```

Comment:
```suggestion
    pub async fn create_table(
        catalog: Arc<dyn MetadataCatalog>,
        options: CreateTableOptions,
    ) -> CatalogResult<Self> {
        let table_name = options.table_name.clone();
        let _table_id = catalog.create_table(options).await?;
        Self::new(&table_name, catalog).await
    }
```

Save cloning the entire options struct for just the table name.